resize/resize_z.m

```diff
@@ -61,48 +64,104 @@
 end
 
-% reshape 2D to match the shape of a 3D field
-if Ny == 1
-    field = reshape(field, [Nx 1 Nz_old]);
-end
-% permute to put extending dimension in the first dimension
-field = permute(field, [3 2 1]);
+if ~strcmp(type_z,'NO_SLIP')
+    
+    % reshape 2D to match the shape of a 3D field
+    if Ny == 1
+        field = reshape(field, [Nx 1 Nz_old]);
+    end
 
-% extension types
-odd_ext    = @(f) [f; -flipud(f)];
-even_ext   = @(f) [f;  flipud(f)];
-period_ext = @(f) [f;  f];
+    % permute to put extending dimension in the first dimension
+    field = permute(field, [3 2 1]);
 
-% double the field
-if strcmp(type_z, 'FOURIER')
-    field = period_ext(field);
-elseif strcmp(field_name, 'w')
-    field = odd_ext(field);
-else
-    field = even_ext(field);
-end
+    % extension types
+    odd_ext    = @(f) [f; -flipud(f)];
+    even_ext   = @(f) [f;  flipud(f)];
+    period_ext = @(f) [f;  f];
 
-% phase shift because grid is cell-centered, and take fft
-dz_new = Lz/Nz_new;
-kzs = fftfreq(2*Nz_old, dz_old);
-phase_corr = exp(-1i*kzs*(dz_old/2-dz_new/2)).';
-field = bsxfun(@times, fft(field), phase_corr);
+    % double the field
+    if strcmp(type_z, 'FOURIER')
+        field = period_ext(field);
+    elseif strcmp(field_name, 'w')
+        field = odd_ext(field);
+    else
+        field = even_ext(field);
+    end
+
+    % phase shift because grid is cell-centered, and take fft
+    dz_new = Lz/Nz_new;
+    kzs = fftfreq(2*Nz_old, dz_old);
+    phase_corr = exp(-1i*kzs*(dz_old/2-dz_new/2)).';
+    field = bsxfun(@times, fft(field), phase_corr);
+
+    % pad or truncate
+    if increase_points
+        % pad with zeros
+        field = [field(1:Nz_old,:,:); zeros(2*Nz_new-2*Nz_old,Ny,Nx); field(end-Nz_old+1:end,:,:)];
+    else
+        % truncate high frequencies
+        field = [field(1:Nz_new,:,:); field(end-Nz_new+1:end,:,:)];
+    end
+
+    % take ifft
+    field = mult*real(ifft(field));
+    field = field(1:Nz_new,:,:);
+    % permute back
+    field = permute(field, [3 2 1]);
+    % return field to 2D if it was
+    if Ny == 1
+        field = reshape(field,[Nx Nz_new]);
+    end
+else 
+    
+
+    %BC must be NO_SLIP. 
+    
+    %Make new Cheb grid
+    [~,z1d_new] = cheb(Nz_new-1);
+    
+    %Make old Cheb grid
+    [~,z1d_old] = cheb(Nz_old-1);
+    
+    %Move to physical space
+    z1d_old = Lz/2*(1 + z1d_old);
+    z1d_new = Lz/2*(1 + z1d_new);
+    
+    %If data is 3D, we must use interp3
+    if Ny ~= 1
+        dx = Lx/Nx;
+        dy = Ly/Ny;
+        field = permute(field, [2 1 3]); %Permute so data agrees with meshgrid
+        %Make old grid
+        [x_old,y_old,z_old] = meshgrid(linspace(dx/2,Lx - dx/2,Nx),...
+            linspace(dy/2,Ly - dy/2,Ny),z1d_old);   
+        %Make new grid
+        [x_new,y_new,z_new] = meshgrid(linspace(dx/2,Lx - dx/2,Nx),...
+            linspace(dy/2,Ly - dy/2,Ny),z1d_new); 
+        %Do interpolation
+        field = interp3(x_old,y_old,z_old,field,x_new,y_new,z_new);
```

Comment:
This works, but to conserve spectral energy, should transform into Chebyshev space and extend there with null high frequency components (as done in Fourier space).

Reply:
Look up Chebfun. It's incredibly powerful and does more than you'll need, but will be a great resource